File: 代码随想录/08_回溯算法/05_电话号码的字母组合/Solution.ts
// 字符串写法
let letterCombinations = (digits: string): string[] => {
    const map = new Map([
        [2, 'abc'], [3, 'def'], [4, 'ghi'], [5, 'jkl'],
        [6, 'mno'], [7, 'pqrs'], [8, 'tuv'], [9, 'wxyz'],
    ]);
    if (!digits.length) return [];
    let res = [];
    const dfs = (path: string, idx: number) => {
        if (idx === digits.length) {
            res.push(path);
            return;
        }
        let letters = map.get(Number(digits[idx]));
        for (let i = 0; i < letters.length; i++) {
            dfs(path + letters.substring(i, i + 1), idx + 1);
        }
    }
    dfs("", 0);
    return res;
};

// 标准的回溯模板
letterCombinations = (digits: string): string[] => {
    const map = new Map([
        [2, 'abc'], [3, 'def'], [4, 'ghi'], [5, 'jkl'],
        [6, 'mno'], [7, 'pqrs'], [8, 'tuv'], [9, 'wxyz'],
    ]);
    if (!digits.length) return [];

    let res = [];
    const dfs = (path: string, idx: number) => {
        if (idx === digits.length) {
            res.push(path);
            return;
        }
        let letters = map.get(Number(digits.charAt(idx)));
        for (let i = 0; i < letters.length; i++) {
            path += letters[i];
            dfs(path, idx + 1);
            path = path.substring(0, path.length - 1);
        }
    }

    dfs("", 0);
    return res;
};

File: 代码随想录/08_回溯算法/05_电话号码的字母组合/Solution.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

// https://leetcode.cn/problems/letter-combinations-of-a-phone-number/
// 字符串截取
class Solution {
    String[] map = new String[] {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    List<String> res = new ArrayList<>();
    String digits;

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) return res;
        this.digits = digits;
        dfs("", 0);
        return res;
    }

    void dfs(String path, int idx) {
        if (idx == digits.length()) {
            res.add(path);
            return;
        }
        String letters = map[digits.charAt(idx) - '0'];
        for (int i = 0; i < letters.length(); i++) {
            dfs(path + letters.substring(i, i + 1), idx + 1);
        }
    }
}

// 标准回溯模板写法 + StringBuilder
class Solution2 {
    Map<Integer, String> map = Map.of(
            2, "abc", 3, "def", 4, "ghi", 5, "jkl",
            6, "mno", 7, "pqrs", 8, "tuv", 9, "wxyz");
    List<String> res = new ArrayList<>();
    String digits;

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) return res;
        this.digits = digits;
        dfs(new StringBuilder(), 0);
        return res;
    }

    void dfs(StringBuilder path, int idx) {
        if (idx == digits.length()) {
            res.add(path.toString());
            return;
        }
        String letters = map.get(digits.charAt(idx) - '0');
        for (int i = 0; i < letters.length(); i++) {
            path.append(letters.charAt(i));
            dfs(path, idx + 1);
            path.deleteCharAt(path.length() - 1);
        }
    }
}